inc/az_ulib_registry_api.h

```diff
@@ -0,0 +1,198 @@
+  /** Total free memory to store registry data In bytes. */
+  size_t free_registry_data;
+} az_ulib_registry_info;
+
+/**
+ * @brief   This function gets the #az_span value associated with the given #az_span key from the
+ * registry.
+ *
+ * This function goes through the registry comparing keys until it finds one that matches the input.
+ *
+ * @param[in]   key                 The #az_span key to look for within the registry.
+ * @param[out]  value               The point to #az_span value corresponding to the input key.
+ *
+ * @pre         Registry shall already be initialized.
+ * @pre         \p key              shall not be `#AZ_SPAN_EMPTY`.
+ * @pre         \p value            shall not be `NULL`.
+ *
+ * @return The #az_result with the result of the registry operations.
+ *      @retval #AZ_OK                        If retrieving a value from the registry was
+ *                                            successful.
+ *      @retval #AZ_ERROR_ITEM_NOT_FOUND      If there are no values that correspond to the
+ *                                            given key within the registry.
+ */
+AZ_NODISCARD az_result az_ulib_registry_try_get_value(az_span key, az_span* value);
+
+/**
+ * @brief   This function adds an #az_span key and an #az_span value into the device registry.
+ *
+ * This function goes through the registry ensuring there are no duplicate elements before
+ * adding a new key value pair to the registry.
+ *
+ * @param[in]   key                 The #az_span key to add to the registry.
+ * @param[in]   value               The #az_span value to add to the registry.
+ *
+ * @pre         Registry shall already be initialized.
+ * @pre         \p key              shall not be `#AZ_SPAN_EMPTY`.
+ * @pre         \p value            shall not be `#AZ_SPAN_EMPTY`.
+ *
+ * @return The #az_result with the result of the registry operations.
+ *      @retval #AZ_OK                            If adding a value to the registry was successful.
+ *      @retval #AZ_ERROR_ULIB_ELEMENT_DUPLICATE  If there is a key within the registry that is the
+ *                                                same as the new key.
+ *      @retval #AZ_ERROR_ULIB_SYSTEM             If the `az_ulib_registry_add` operation failed on
+ *                                                the system level.
+ *      @retval #AZ_ERROR_ULIB_BUSY               If the resources necessary for the
+ *                                                `az_ulib_registry_add` operation are busy.
+ *      @retval #AZ_ERROR_OUT_OF_MEMORY           If the flash space for `az_ulib_registry_add`
+ *                                                is not enough for a new registry entry.
+ */
+AZ_NODISCARD az_result az_ulib_registry_add(az_span key, az_span value);
+
+/**
+ * @brief   This function removes an #az_span key and its corresponding #az_span value from the
+ * device registry.
+ *
+ * This function goes through the registry and marks the node with the given key value pair as
```

Comment:
Maybe mention here that the memory associated with value is not truly cleared until defrag occurs.

Reply:
I prefer to have this information when we have the defrag function, and have this comment in the defrag function not here.

Reply:
#Wontfix